rfcs/remote_channel.md

```diff
@@ -0,0 +1,616 @@
+
+With the stash-based approach, it's important to clean up the stash
+once no channels remain. Otherwise we end up leaking queues. The
+general approach is to explicitly store a refcount as part of the
+stash value. Each socket that connects increments the refcount, and
+decrements it once the socket is closed. Access to the stash is
+protected by a lock, so only one socket may touch the refcount at a
+time. In practice the fact that there can only be one `RecvChannel`
+per queue means that we can just count `SendChannel` sockets and
+enforce an at-most-one rule for `RecvChannel`. The underlying queue is
+deleted when no socket connections remain. This does mean that if we
+were to e.g. put a message on a queue, then delete all the channels
+before forwarding the message, and then use the UUID to create a new
+`RecvChannel` the queue would be deleted and the message lost. One can
+imagine this scenario happening when navigating a remote context which
+has a `RecvChannel` tied to the UUID in the URL, if the `SendChannel`
+for the context isn't kept alive. Another possibility is to only
+delete the queue if it's empty, and register an explicit completion
+callback in the test that deletes any queue with a UUID known to the
+test context. Of course this doesn't help if a channel is created in a
+non-test context (e.g. for messaging between multiple remotes).
+
+## Example
+
+test.html
+
+```html
+<!doctype html>
+<title>executeScript example</title>
+<script src="/resources/testharness.js">
+<script src="/resources/testharnessreport.js">
+<script src="/resources/channel.js">
+
+<script>
+promise_test(async t => {
+  let remote = new RemoteWindow();
+  window.open(`child.html?uuid=${remote.uuid}`, "_blank", "noopener");
+  let result = await remote.executeScript(id => {
+    return document.getElementById(id).textContent;
+  }, "test");
+  assert_equals("result", PASS);
+});
+</script>
+```
+
+child.html
+
+```html
+<p id="nottest">FAIL</p>
+<p id="test">PASS</p>
+<script>
+start_window();
+</script>
+```
+
+## Possible Future Additions
+
+The primitives here could be integrated more completely with
+testharness.js. For example we could use a `RemoteWindow` as a source
+of tests in `fetch_tests_from_window`. Alternatively, or in addition
+to that, we could integrate asserts with script execution, so that a
+remote context could include a minimal testharness.js and enable
+something like:
+
+```js
+promise_test(t => {
+  let r = new RemoteWindow();
+  window.open(`file.html?uuid=${r.uuid}`, "_blank", "noopener");
+  await r.executeScript(() => {
+    assert_equals(window.opener, null)
+  });
+});
+```
+
+Currently this is possible if `file.html` is put in `resources` or
+`support` directory so it can't be detected as a test. But it would
+perhaps be better to have a minimal subset of testharness available
+without providing the API surface that only makes sense in a test
+window.
+
+It may be possible in the future to replace the backend with a
```

Comment:
For further integration with WebDriver, testdriver etc., discussion about benefits and risks would be helpful (not necessarily now though, given that this is "Possible Future Additions").

One one hand ("Risk" side), I basically prefer not integrating the framework too much to WebDriver/test_driver etc., to make the framework work with minimum dependencies and work with bare browsers, and not to make debugging harder. We might want to `JSON.stringify()` instead of WebDriver BiDi format if we want to reduce complexity around serialization, given that the current tests considered so far don't need WebDriver BiDi capability beyond `JSON.stringify()`.
So I'm interested in hearing more about the benefits (from the point of view of test writers and WPT infra maintainers) when we consider the further integration.

Reply:
Yes, future additions here would be subject to the same RFC process. So the idea was just to sketch out some possibilites in case people felt there were specific parts that should be given additional consideration in the initial design, or would be high value to work on sooner.

In terms of the serialization format, we are already using extra complexity beyond JSON-stringify when serializing a `SendChannel` for message responses. We obviously could special case that, but in general I think using plain JSON as a serialization format for data where we might want to support non-JSON types is a bad idea (e.g. current WebDriver mostly uses plain JSON but also has super-hacky support for serializing Element objects which doesn't generalise to other types and is generally a mess). Aiming for a featureset closer to structured clone seems more future proof, but structured clone doesn't actually define a serialization. So, given that WebDriver BiDi has very similar requirements, and actually defines the wire format, it seems sensible to use that as a reference point. The fact that, going forward, we could use WebDriver BiDi instead of the stash as the backend in the wptrunner implementation is good, but it's not the primary motivation here.